### DISCUSSION TITLE
Looking to improve on the target sensors

### DISCUSSION THREAD

**Discussion by Rhysgh** [BottlecapDave/HomeAssistant-OctopusEnergy#643] [Ideas]
Hi, im making use of the update target sensor features and I'm enjoying the results but I feel like I want to try and make a change.. 

I have looked into /config/custom_components/octopus_energy/config/target_rates.py and see some handeling at lines 94-115 but I thought I would ask here before making any changes. 

attached is a photo that helps explain my issue 
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/57257743/34c73137-cac6-490c-89f2-9385a1bfe3ea)

you see the blue marks indicate when the target sensor will be active.. as you can see there is 3 slots that are yellow at 20:00, 05:30 and 12:20 where as there are 3 slots at 21:30..22:30 that would be more beneficial. 

this is a problem now because its 20:00 so I cant set the end time of the target sensor to 22:30 because then ill be told there isn't enough hours (when there is because I'm talking about tomorrow) see the issue? I'm looking to evaluate based on all available data rather than within a time period.. whats your thoughts?

**Comment by Rhysgh** [comment]
I found this in /config/custom_components/octopus_energy/target_rates/__init__.py

``` # If our start date has passed, reset it to current_date to avoid picking a slot in the past
  if (is_rolling_target == True and target_start < current_date and current_date < target_end):
    _LOGGER.debug(f'Rolling target and {target_start} is in the past. Setting start to {current_date}')
    target_start = current_date

  # If our start and end are both in the past, then look to the next day
  if (target_start < current_date and target_end < current_date):
    target_start = target_start + timedelta(days=1)
    target_end = target_end + timedelta(days=1)``` 

It could be here maybe.. what I would like to try is is_time_frame_long_enough == false then add 1 day to the end time

**Reply by BottlecapDave** [reply]
I'm not sure of your full config, but if you set the from/to time 20:00 to 20:00, turned off rolling target and set it to continuous then it would either start at 21:30 or it would end at 06:00 depending on your target hours. But it looks like your have intermittent which is why it's picking hours not grouped together.

The rolling targets should probably behave how you're describing. Rolling targets might not be what you're after, but without knowing the full context it's a bit hard to say otherwise.

**Reply by Rhysgh** [reply]
this is my setup, 

```
Rolling target
false
Kind
target_rate
Name
import_target
Hours
7.5
Type
Intermittent
Mpan
2100040663425
Account ID
A-48DD78BA
Last rates
false
Target invert target rates
false
Start time
20:00
End time
20:00
Is target export
false
Target times
- value_inc_vat: 0.154035
start: '2023-12-18T20:00:00+00:00'
end: '2023-12-18T20:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.154035
start: '2023-12-18T22:30:00+00:00'
end: '2023-12-18T23:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.14763
start: '2023-12-18T23:00:00+00:00'
end: '2023-12-18T23:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.13398
start: '2023-12-18T23:30:00+00:00'
end: '2023-12-19T00:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.13839
start: '2023-12-19T00:00:00+00:00'
end: '2023-12-19T00:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.14763
start: '2023-12-19T01:00:00+00:00'
end: '2023-12-19T01:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.14091
start: '2023-12-19T01:30:00+00:00'
end: '2023-12-19T02:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.14763
start: '2023-12-19T02:00:00+00:00'
end: '2023-12-19T02:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.123585
start: '2023-12-19T02:30:00+00:00'
end: '2023-12-19T03:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.14763
start: '2023-12-19T03:00:00+00:00'
end: '2023-12-19T03:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.13167
start: '2023-12-19T03:30:00+00:00'
end: '2023-12-19T04:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.14763
start: '2023-12-19T04:00:00+00:00'
end: '2023-12-19T04:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.137235
start: '2023-12-19T04:30:00+00:00'
end: '2023-12-19T05:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.15015
start: '2023-12-19T05:30:00+00:00'
end: '2023-12-19T06:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.15456
start: '2023-12-19T12:30:00+00:00'
end: '2023-12-19T13:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
Target times last evaluated
18 December 2023 at 20:00:04
Overall average cost
0.14
Overall min cost
0.12
Overall max cost
0.15
Current duration in hours
0
Current average cost
Unknown
Current min cost
Unknown
Current max cost
Unknown
Next time
18 December 2023 at 22:30:00
Next duration in hours
2
Next average cost
0.14
Next min cost
0.13
Next max cost
0.15
Last evaluated
18 December 2023 at 20:43:54
```

I'm changing it every hour I want to evaluate all available rates so until 22:30 tomorrow but I can't set 22:30 as the end time because the sensor assumes I mean today at 22:30 as that time isn't in the past yet .. get me?

**Reply by Rhysgh** [reply]
Please don't get confused with intermittent, i do want that, that's not my issue it's the fact that the green slots at 21:30 onwards are ignored that's my issue.

**Reply by BottlecapDave** [reply]
Ok, the reason for this is that we need to keep track of when the sensor comes on to ensure it comes on once per 24 hour periods. If the available rates keep changing then there's a chance the sensor would never come on because it would keep moving forward to the next set. This goes out the window if you update the config manually. 

So in your scenario if your window was between 00:00 and 00:00, it would pick most of the green rates. 

Of course you're welcome to fork the integration and make your own local changes to meet your needs, but doing so at your own risk.

**Reply by Rhysgh** [reply]
Right yea, i see your point.. let me sleep on it.

**Reply by Rhysgh** [reply]
I would value your opinion. for example, at the moment I will try and update the target to (its 21:40 while I'm doing this)

```{"hoursNeededToCharge":7,"targettimestart":"21:30","targettimeend":"22:30"}```

ill get the error ~"not enough hours" ( I mean 22:30 tomorrow but we understand why that's not done) so in place ill post 

```{"hoursNeededToCharge":7,"targettimestart":"21:30","targettimeend":"21:30"}```

the hours being new Date() , I figure this is giving me a rolling 24 hours as a worst case scenario if my desired times are not allowed. 

does that sound like a reasonable comprise to you?

**Reply by BottlecapDave** [reply]
Rolling times should work as long as they're the same time. If they're different (e.g 10:00 to 16:00) then you might hit issues when it comes to the cross over period as outlined in the docs. Ill try and think if theres any way of expanding with additional config that could get you where you want, but I'm already a little concerned with the amount of configuration available.

**Reply by Rhysgh** [reply]
If you would like I’ll share with you my full config, it may help you better understand what I’m trying to achieve (achieving actually) these are small issues that I’m striving for perfection over

**Reply by Rhysgh** [reply]
just to expand on this today.. see not 16:00-23:00 times ( cos we have an array until 23:00) I get this 

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/57257743/7ab0b9e8-25a1-4ee0-8dda-3a5cc48f7bf0)

But at 16:00-16:00 I get 

 
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/57257743/2e21bd69-7816-4b93-a8f1-46aa646f6f66)


those two blue ones on Thu must be a bug with the card.. 

here is the sensor

```
Rolling target
false
Kind
target_rate
Name
import_target
Hours
7
Type
Intermittent
Mpan
2100040663425
Account ID
A-48DD78BA
Last rates
false
Target invert target rates
false
Start time
16:00
End time
16:00
Is target export
false
Rates incomplete
false
Target times
- value_inc_vat: 0.02772
start: '2023-12-20T22:30:00+00:00'
end: '2023-12-20T23:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.010395
start: '2023-12-20T23:30:00+00:00'
end: '2023-12-21T00:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.019215
start: '2023-12-21T00:00:00+00:00'
end: '2023-12-21T00:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.03465
start: '2023-12-21T00:30:00+00:00'
end: '2023-12-21T01:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.014595
start: '2023-12-21T01:00:00+00:00'
end: '2023-12-21T01:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0
start: '2023-12-21T01:30:00+00:00'
end: '2023-12-21T02:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: -0.01407
start: '2023-12-21T02:00:00+00:00'
end: '2023-12-21T02:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: -0.03927
start: '2023-12-21T02:30:00+00:00'
end: '2023-12-21T03:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: -0.01806
start: '2023-12-21T03:00:00+00:00'
end: '2023-12-21T03:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: -0.0231
start: '2023-12-21T03:30:00+00:00'
end: '2023-12-21T04:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: -0.010395
start: '2023-12-21T04:00:00+00:00'
end: '2023-12-21T04:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: -0.01827
start: '2023-12-21T04:30:00+00:00'
end: '2023-12-21T05:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.00441
start: '2023-12-21T05:00:00+00:00'
end: '2023-12-21T05:30:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
- value_inc_vat: 0.02982
start: '2023-12-21T05:30:00+00:00'
end: '2023-12-21T06:00:00+00:00'
tariff_code: E-1R-AGILE-FLEX-22-11-25-K
is_capped: false
Target times last evaluated
20 December 2023 at 16:06:03
Overall average cost
0
Overall min cost
-0.04
Overall max cost
0.03
Current duration in hours
0
Current average cost
Unknown
Current min cost
Unknown
Current max cost
Unknown
Next time
20 December 2023 at 22:30:00
Next duration in hours
0.5
Next average cost
0.03
Next min cost
0.03
Next max cost
0.03
Last evaluated
20 December 2023 at 16:10:16
```
